### PR TITLE
Separated application in two modules: app and commons [Issue #73]

### PR DESCRIPTION
**Pull request by jusleg**
**Issue #73**

### WHAT kind of change does this PR introduce?
This is a large PR. It breaks the application into two different modules: `app` for the phone part and `commons` for the code that will be reused by the phone and the wear app. This will make it easier to start the android wear app and will allow for code reuse. the mobile app has access to what is in `commons` but not the other way around. This means that when we make the android wear app, the wear app will not contain the layout or values that are only used by the phone app.

* App is now in two modules
* the CI runs the tests for both modules
* Broke dependency of the Calculator activity in CalculatorImpl by using a fake object. A mock should be done in the future for both CalculatorImpl and MoneyCalculatorImpl
* Fixed warning in the console while building

screen shot 2018-03-09 at 9 48 33 pm

### HOW is this accomplished?
Moving a lot of files, updating the import paths and cleaning the gradle configs.

### Checklist
- [x] Docs have been added / updated to reflect the changes
- [x] I have reviewed and tested the changes :white_check_mark:
- [x] I have squashed my commits to have a reasonable amount of commits
- [x] All of the code I have written adhere to the coding conventions outlined in the coding conventions section of the wiki

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -24,5 +24,8 @@ before_script:
   - android-wait-for-emulator
   - adb shell input keyevent 82 &
 
+after_success:
+  - bash <(curl -s https://codecov.io/bash)
```
Comment (almiche) [inline .travis.yml]: Nice!

File: app/build.gradle [hunk 2]
```diff
@@ -6,9 +12,14 @@ android {
...
 
+    compileOptions {
+        kotlinOptions.freeCompilerArgs += ['-module-name', "$PUBLISH_GROUP_ID.$PUBLISH_ARTIFACT_ID"]
+    }
+
+
     defaultConfig {
         applicationId "com.simplemobiletools.calculator"
-        minSdkVersion 16
+        minSdkVersion 22
```
Comment (almiche) [inline app/build.gradle]: On to the future !

File: app/build.gradle [hunk 3]
```diff
@@ -29,6 +40,17 @@ android {
...
 
+    packagingOptions {
+        exclude 'META-INF/DEPENDENCIES'
+        exclude 'META-INF/LICENSE'
+        exclude 'META-INF/LICENSE.txt'
+        exclude 'META-INF/license.txt'
+        exclude 'META-INF/NOTICE'
+        exclude 'META-INF/NOTICE.txt'
+        exclude 'META-INF/notice.txt'
+        exclude 'META-INF/ASL2.0'
```
Comment (almiche) [inline app/build.gradle]: Always good to exclude these

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -64,17 +55,5 @@
             android:name="com.simplemobiletools.commons.activities.CustomizationActivity"
             android:label="@string/customize_colors"
             android:parentActivityName=".activities.SettingsActivity"/>
-
```
Comment (almiche) [inline app/src/main/AndroidManifest.xml]: Bout time we deleted that widget!

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 5]
```diff
@@ -2,29 +2,24 @@ package com.simplemobiletools.calculator.activities
...
 import android.support.v4.view.ViewCompat
-import android.util.Log
 import android.view.View
 import android.widget.AdapterView
 import android.widget.ListView
 import com.simplemobiletools.calculator.R
-import com.simplemobiletools.calculator.extensions.config
-import com.simplemobiletools.calculator.extensions.updateViewColors
-import com.simplemobiletools.calculator.helpers.Calculator
+import com.simpletools.calculator.commons.extensions.config
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: What is this change ?

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -6,12 +6,10 @@
     <string name="vibrate_on_button_press">Vibrate on button press</string>
 
     <!-- Strings displayed only on Google Playstore. Optional, but good to have -->
-    <string name="app_short_description">A calculator with the basic functions and a customizable widget.</string>
+    <string name="app_short_description">A calculator with the basic functions.</string>
```
Comment (almiche) [inline app/src/main/res/values/strings.xml]: Good call

File: build.gradle [hunk 7]
```diff
@@ -11,12 +12,13 @@ buildscript {
     dependencies {
         classpath 'com.android.tools.build:gradle:3.0.1'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
-
+        classpath 'com.dicedmelon.gradle:jacoco-android:0.1.2'
```
Comment (almiche) [inline build.gradle]: What is jacoco?

File: commons/src/test/java/com/simpletools/calculator/commons/CalculatorImplTest.kt [hunk 8]
```diff
@@ -1,26 +1,25 @@
...
 class CalculatorImplTest {
-    private lateinit var activity: MainActivity
+    private lateinit var activity: Calculator
     private lateinit var calc: CalculatorImpl
 
     @Before
     fun setUp() {
-        activity = Robolectric.setupActivity(MainActivity::class.java)
-        calc = CalculatorImpl(activity, activity.applicationContext)
+        activity = FakeCalculator()
```
Comment (almiche) [inline commons/src/test/java/com/simpletools/calculator/commons/CalculatorImplTest.kt]: Great addition!

File: commons/src/test/java/com/simpletools/calculator/commons/FakeCalculatorActivity.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package com.simpletools.calculator.commons
+
+import android.content.Context
+import com.simpletools.calculator.commons.activities.SimpleActivity
+import com.simpletools.calculator.commons.helpers.Calculator
+
+class FakeCalculator: SimpleActivity(), Calculator {
```
Comment (almiche) [inline commons/src/test/java/com/simpletools/calculator/commons/FakeCalculatorActivity.kt]: What's the difference between this class and just the regular Calculator ?

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 5 above)
Comment (jusleg) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: the config has moved to the commons module but it is used in the app module for the phone. The import path is therefore different

File: commons/src/test/java/com/simpletools/calculator/commons/FakeCalculatorActivity.kt (hunk 9 above)
Comment (jusleg) [reply commons/src/test/java/com/simpletools/calculator/commons/FakeCalculatorActivity.kt]: no activity behind it. basically I kept the method signatures to make it behave like a mock

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt [hunk 10]
```diff
@@ -1,11 +1,12 @@
...
-import com.simplemobiletools.calculator.R
-import com.simplemobiletools.calculator.operation.TaxOperation
-import com.simplemobiletools.calculator.operation.TipOperation
+import com.simpletools.calculator.commons.helpers.Calculator
+import com.simpletools.calculator.commons.R
+import com.simpletools.calculator.commons.operations.TaxOperation
+import com.simpletools.calculator.commons.operations.TipOperation
+import com.simpletools.calculator.commons.helpers.Formatter
 
-class MoneyCalculatorImpl(calculator: Calculator,taxCalculator: TaxCalculator, val context: Context) {
```
Comment (0matleb2) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt]: Wondering why this Context parameter is being removed in a bunch of places. Was it unnecessary?

File: commons/src/test/java/com/simpletools/calculator/commons/CalculatorImplTest.kt [hunk 11]
```diff
@@ -120,9 +119,7 @@ class CalculatorImplTest {
         calc.handleOperation("plus")
         calc.addDigit(2)
         calc.handleClear()
-        assertEquals(activity.btn_clear.text,"AC")
```
Comment (0matleb2) [inline commons/src/test/java/com/simpletools/calculator/commons/CalculatorImplTest.kt]: Good, don't want to test too many things in one test.

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt (hunk 10 above)
Comment (jusleg) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt]: Was mostly there for the widget. Made a dependency for the test harness. Couldn't find a good way to break it while keeping the context. I opted to removed the widget as it was legacy and all the other instances of calculator did not need it. Simpler code, lower coupling.

File: commons/src/test/java/com/simpletools/calculator/commons/CalculatorImplTest.kt (hunk 11 above)
Comment (jusleg) [reply commons/src/test/java/com/simpletools/calculator/commons/CalculatorImplTest.kt]: I didn't want to test UI stuff in a unit test also, the behavior is already tested. it is not the value of a string that will confirm the behavior.

File: build.gradle [hunk 12]
```diff
@@ -1,7 +1,7 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
-    ext.kotlin_version = '1.2.0'
+    ext.kotlin_version = '1.2.30'
```
[the remark this answers is not in the document]
Comment (jusleg) [reply build.gradle]: Always moving forward!

File: app/build.gradle [hunk 13]
```diff
@@ -59,6 +81,8 @@ dependencies {
     androidTestImplementation 'com.android.support.test:runner:1.0.1'
     androidTestImplementation 'com.android.support.test:rules:1.0.1'
     androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.1'
+    compile project(':commons')
```
[the remark this answers is not in the document]
Comment (jusleg) [reply app/build.gradle]: I'll look into that. I'll keep you posted on what I can do for that

[the remark this answers is not in the document]
Comment (jusleg) [reply app/build.gradle]: switched to implementation!

**Comment by almiche** [comment]
Styling Review: All looks well and conforms to our guidelines!

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt [hunk 14]
```diff
@@ -97,7 +98,7 @@ class MoneyCalculatorImpl(calculator: Calculator,taxCalculator: TaxCalculator, v
     }
 
     fun performTaxing(location:String){
-        overwriteNumber(TaxOperation(getResult(),location).getResult())
```
Comment (izconcept) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt]: Good styling catch!

File: app/src/test/java/com/simplemobiletools/calculator/MoneyCalculatorImplTest.kt [hunk 15]
```diff
@@ -19,7 +19,7 @@ class MoneyCalculatorImplTest {
     @Before
     fun setUp() {
         activity = Robolectric.setupActivity(MoneyActivity::class.java)
-        calc = MoneyCalculatorImpl(activity,activity, activity.applicationContext)
+        calc = MoneyCalculatorImpl(activity,activity)
```
Comment (izconcept) [inline app/src/test/java/com/simplemobiletools/calculator/MoneyCalculatorImplTest.kt]: **Style critique: ** Missing a space after first parameter comma.

Comment (jusleg) [reply app/src/test/java/com/simplemobiletools/calculator/MoneyCalculatorImplTest.kt]: updated

**Pull request merged**